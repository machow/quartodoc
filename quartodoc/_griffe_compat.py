# flake8: noqa

try:
    from griffe import GriffeLoader
    from griffe import ModulesCollection, LinesCollection

    import _griffe.models as dataclasses
    import _griffe.docstrings.models as docstrings
    import _griffe.expressions as expressions

    from griffe import Parser, parse, parse_numpy
    from griffe import AliasResolutionError

    from griffe import load_extensions
except ImportError:
    from griffe.loader import GriffeLoader
    from griffe.collections import ModulesCollection, LinesCollection

    import griffe.dataclasses as dataclasses
    import griffe.docstrings.dataclasses as docstrings
    import griffe.expressions as expressions

    from griffe.docstrings.parsers import Parser, parse
    from griffe.exceptions import AliasResolutionError

    from griffe import load_extensions
