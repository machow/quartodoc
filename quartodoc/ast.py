from __future__ import annotations

import warnings

from enum import Enum
from dataclasses import dataclass
from griffe.docstrings import dataclasses as ds
from griffe import dataclasses as dc
from plum import dispatch
from typing import Union
from pydantic import BaseModel  # for previewing


# Transform and patched-in classes ============================================
# TODO: annotate transform return types. make sure subtypes inherit from correct
# griffe base objects.
# TODO: it seems like transform should happen on the root, not individual elements.


def transform(el):
    """Return a more specific docstring element, or simply return the original one."""

    if isinstance(el, tuple):
        try:
            return tuple_to_data(el)
        except ValueError:
            pass

    # patch a list of docstring sections. note that this has to happen on the
    # list, since we replace single nodes on the tree (the list is the node).
    elif isinstance(el, list) and len(el) and isinstance(el[0], ds.DocstringSection):
        return _DocstringSectionPatched.transform_all(el)

    return el


# Patch DocstringSection ------------------------------------------------------


class DocstringSectionKindPatched(Enum):
    see_also = "see also"
    notes = "notes"
    warnings = "warnings"


class _DocstringSectionPatched(ds.DocstringSection):
    _registry: "dict[Enum, _DocstringSectionPatched]" = {}

    def __init__(self, value: str, title: "str | None" = None):
        super().__init__(title)
        self.value = value

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

        if cls.kind.value in cls._registry:
            raise KeyError(f"A section for kind {cls.kind} already exists")

        cls._registry[cls.kind.value] = cls

    @staticmethod
    def split_sections(text) -> list[tuple[str, str]]:
        """Return tuples of (title, body) for all numpydoc style sections in the text.

        Note that this function does not check the value of the section header,
        only that it is a header on one line, with dashed on the next.
        """
        import re

        comp = re.compile(r"^([\S \t]+)\n-+$\n?", re.MULTILINE)

        crnt_match = comp.search(text)
        crnt_pos = 0

        # This loop takes a match, then attempts to look ahead at the next one,
        # in order to find the body of text between multiple sections.
        results = []
        while crnt_match is not None:
            next_pos = crnt_pos + crnt_match.end()
            substr = text[next_pos:]
            next_match = comp.search(substr)

            title = crnt_match.groups()[0]
            body = substr if next_match is None else substr[: next_match.start()]

            results.append((title, body))

            crnt_match, crnt_pos = next_match, next_pos

        return results

    @classmethod
    def transform(cls, el: ds.DocstringSection) -> list[ds.DocstringSection]:
        """Attempt to cast DocstringSection element to more specific section type.

        Note that this is meant to patch cases where the general DocstringSectionText
        class represents a section like See Also, etc..
        """

        if not isinstance(el, ds.DocstringSectionText):
            return [el]

        splits = cls.split_sections(el.value)
        results = []
        for title, body in splits:
            sub_cls = cls._registry.get(title.lower(), ds.DocstringSectionText)

            # note that griffe currently doesn't store the title anywhere,
            # but we add the exact title here, so we can be flexible about the
            # sections we parse (e.g. Note instead of Notes) in the future.
            results.append(sub_cls(body, title))

        return results or [el]

    @classmethod
    def transform_all(cls, el: list[ds.DocstringSection]) -> list[ds.DocstringSection]:
        return sum(map(cls.transform, el), [])


class DocstringSectionSeeAlso(_DocstringSectionPatched):
    kind = DocstringSectionKindPatched.see_also


class DocstringSectionNotes(_DocstringSectionPatched):
    kind = DocstringSectionKindPatched.notes


class DocstringSectionWarnings(_DocstringSectionPatched):
    kind = DocstringSectionKindPatched.warnings


# Patch Example elements ------------------------------------------------------


@dataclass
class ExampleCode:
    value: str


@dataclass
class ExampleText:
    value: str


def tuple_to_data(el: "tuple[ds.DocstringSectionKind, str]"):
    """Re-format funky tuple setup in example section to be a class."""
    assert len(el) == 2

    kind, value = el
    if kind.value == "examples":
        return ExampleCode(value)
    elif kind.value == "text":
        return ExampleText(value)

    raise ValueError(f"Unsupported first element in tuple: {kind}")


# Tree previewer ==============================================================


@dispatch
def fields(el: BaseModel):
    # TODO: this is the only quartodoc specific code.
    # pydantic seems to copy MISSING() when it's a default, so we can't
    # whether a MISSING() is the default MISSING(). Instead, we'll just
    # use isinstance for this particular class
    from .layout import MISSING

    # return fields whose values were not set to the default
    field_defaults = {mf.name: mf.default for mf in el.__fields__.values()}
    return [
        k for k, v in el if field_defaults[k] is not v if not isinstance(v, MISSING)
    ]


@dispatch
def fields(el: dc.Object):
    options = [
        "name",
        "canonical_path",
        "classes",
        "parameters",
        "members",
        "functions",
        "docstring",
    ]
    return [opt for opt in options if hasattr(el, opt)]


@dispatch
def fields(el: dc.ObjectAliasMixin):
    try:
        return fields(el.target)
    except dc.AliasResolutionError:
        warnings.warn(
            f"Could not resolve Alias target `{el.target_path}`."
            " This often occurs because the module was not loaded (e.g. it is a"
            " package outside of your package)."
        )
        return ["name", "target_path"]


@dispatch
def fields(el: dc.Function):
    return ["name", "annotation", "parameters", "docstring"]


@dispatch
def fields(el: dc.Attribute):
    return ["name", "annotation"]


@dispatch
def fields(el: dc.Docstring):
    return ["parser", "parsed"]


@dispatch
def fields(el: ds.DocstringSection):
    return ["kind", "title", "value"]


@dispatch
def fields(el: ds.DocstringParameter):
    return ["annotation", "default", "description", "name", "value"]


@dispatch
def fields(el: ds.DocstringElement):
    return ["annotation", "description"]


@dispatch
def fields(el: ds.DocstringNamedElement):
    return ["name", "annotation", "description"]


@dispatch
def fields(el: dc.Parameter):
    return ["annotation", "kind", "name", "default"]


@dispatch
def fields(el: dict):
    return list(el.keys())


@dispatch
def fields(el: Union[list, dc.Parameters]):
    return list(range(len(el)))


@dispatch
def fields(el: object):
    if isinstance(el, (ExampleCode, ExampleText)):
        from dataclasses import fields as _fields

        return [field.name for field in _fields(el)]

    return None


class Formatter:
    n_spaces = 3
    icon_block = "█─"
    icon_pipe = "├─"
    icon_endpipe = "└─"
    icon_connector = "│ "
    string_truncate_mark = " ..."

    def __init__(self, string_max_length: int = 50, max_depth=999, compact=False):
        self.string_max_length = string_max_length
        self.max_depth = max_depth
        self.compact = compact

    def format(self, call, depth=0, pad=0):
        """Return a Symbolic or Call back as a nice tree, with boxes for nodes."""

        call = transform(call)

        crnt_fields = fields(call)

        if crnt_fields is None:
            str_repr = repr(call)
            if len(str_repr) > self.string_max_length:
                return str_repr[: self.string_max_length] + self.string_truncate_mark

            return str_repr

        call_str = self.icon_block + call.__class__.__name__

        # short-circuit for max depth ----
        if depth >= self.max_depth:
            return call_str + self.string_truncate_mark

        # format arguments ----
        fields_str = []
        for name in crnt_fields:
            val = self.get_field(call, name)

            # either align subfields with the end of the name, or put the node
            # on a newline, so it doesn't have to be so indented.
            if self.compact:
                sub_pad = pad
                linebreak = "\n" if fields(val) else ""
            else:
                sub_pad = len(str(name)) + self.n_spaces
                linebreak = ""

            # do formatting
            formatted_val = self.format(val, depth + 1, pad=sub_pad)
            fields_str.append(f"{name} = {linebreak}{formatted_val}")

        padded = []
        for ii, entry in enumerate(fields_str):
            is_final = ii == len(fields_str) - 1

            chunk = self.fmt_pipe(entry, is_final=is_final, pad=pad)
            padded.append(chunk)

        return "".join([call_str, *padded])

    def get_field(self, obj, k):
        if isinstance(obj, (dict, list, dc.Parameters)):
            return obj[k]

        return getattr(obj, k)

    def fmt_pipe(self, x, is_final=False, pad=0):
        if not is_final:
            connector = self.icon_connector if not is_final else "  "
            prefix = self.icon_pipe
        else:
            connector = "  "
            prefix = self.icon_endpipe

        connector = "\n" + " " * pad + connector
        prefix = "\n" + " " * pad + prefix
        # NOTE: because visiting is depth first, this is essentially prepending
        # the text to the left.
        return prefix + connector.join(x.splitlines())


def preview(
    ast: "dc.Object | ds.Docstring | object",
    max_depth=999,
    compact=False,
    as_string: bool = False,
):
    """Print a friendly representation of a griffe object (e.g. function, docstring)

    Examples
    --------

    >>> from quartodoc import get_object
    >>> obj = get_object("quartodoc", "get_object")

    >>> preview(obj.docstring.parsed)
     ...

    >>> preview(obj)
     ...

    """

    res = Formatter(max_depth=max_depth, compact=compact).format(ast)

    if as_string:
        return res

    print(res)
