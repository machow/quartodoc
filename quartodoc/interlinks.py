"""The interlinks module fully specifies the behavior of an interlinks filter.

Note that this module largely exists for two reasons:

* Allow people to explore / debug interlinks in python.
* Provide a reference implementation for the lua interlinks filter.

See quartodoc.tests.test_interlinks for its implementation, and the fully
loaded specification.
"""


from __future__ import annotations

import os
import itertools
import json
import warnings
import yaml

from pydantic import BaseModel, Field
from dataclasses import dataclass
from pathlib import Path
from typing import Literal, Annotated, Union, Optional


ENV_PROJECT_ROOT: str = "QUARTO_PROJECT_ROOT"

# Errors -----------------------------------------------------------------------


class RefSyntaxError(Exception):
    """An error parsing an interlinks reference."""


class InvLookupError(Exception):
    """An error looking up an entry from inventory files."""


# Utility functions -----------------------------------------------------------


def get_path_to_root():
    # In lua filters you can use quarto.project.offset
    return Path(os.environ[ENV_PROJECT_ROOT])


def parse_rst_style_ref(full_text: str):
    """
    Returns
    -------
    tuple
        The parsed title (None if no title specified), and corresponding reference.
    """

    import re

    m = re.match(r"(?P<text>.+?)\<(?P<ref>[a-zA-Z\.\-: _]+)\>", full_text)
    if m is None:
        # TODO: print a warning or something
        return full_text, None

    text, ref = m.groups()

    return ref, text


def parse_md_style_link(full_text: str):
    import re

    m = re.match(r"\[(?P<text>.*?)\]\((?P<ref>.*?)\)", full_text)

    if m is None:
        raise Exception()

    text, ref = m.groups()

    return ref, text


# Dataclasses representing pandoc elements ------------------------------------
# These classes are used to help indicate what elements the Interlinks class
# would return in a pandoc filter.


class Link(BaseModel):
    """Indicates a pandoc Link element."""

    kind: Literal["Link"] = "Link"
    content: str
    url: str


class Code(BaseModel):
    """Indicates a pandoc Code element."""

    kind: Literal["Code"] = "Code"
    content: str


class Unchanged(BaseModel):
    """Marker class for content that a function no-ops.

    The main purpose of this class is to indicate when a pandoc filter might
    return the original content element.
    """

    kind: Literal["Unchanged"] = "Unchanged"
    content: str


class TestSpecEntry(BaseModel):
    input: str
    output_text: Optional[str] = None
    output_link: Optional[str] = None
    output_element: Optional[
        Annotated[Union[Link, Code, Unchanged], Field(discriminator="kind")]
    ] = None
    error: Optional[str] = None
    warning: Optional[str] = None


class TestSpec(BaseModel):
    __root__: list[TestSpecEntry]


# Reference syntax ------------------------------------------------------------
# note that the classes above were made pydantic models so we could serialize
# them from json. We could make these ones pydantic too, but there is not a
# ton of benefit here.


@dataclass
class Ref:
    """Represent a sphinx-style reference.

    These have this format
        :external+<invname>:<domain>:<role>:`<target>`

    """

    target: "str"
    role: "None | str" = None
    domain: "None | str" = None
    invname: "None | str" = None

    external: bool = False

    @classmethod
    def from_string(cls, ref: str):
        if not (ref.startswith(":") or ref.startswith("`")):
            raise RefSyntaxError(
                'Ref must start with ":" or "`".\n' f"Received ref string: {ref}"
            )

        if not ref.endswith("`"):
            raise RefSyntaxError(
                'Ref must end with "`"\n' f"Received ref string: {ref}"
            )

        # Note that optional options after :external: go right-to-left.
        # e.g. :role:`target`
        # e.g. :external:role:`target`
        # e.g. :external:domain:role:`target`

        kwargs = {}

        # TODO: user may have omitted the starting `
        params, kwargs["target"], _ = ref.rsplit("`", 2)

        if params != "":
            if ref.startswith(":external"):
                external, *parts = params.lstrip(":").rstrip(":").split(":")

                kwargs["external"] = True
                if "+" in external:
                    kwargs["invname"] = external.split("+")[-1]
                else:
                    kwargs["invname"] = None

            else:
                kwargs["invname"] = None
                parts = params.lstrip(":").rstrip(":").split(":")

            kwargs.update(zip(["role", "domain"], reversed(parts)))

        return cls(**kwargs)


# Hold all inventory items in a singleton -------------------------------------


@dataclass
class EnhancedItem:
    # these are defined in the quarto config
    inv_name: str
    inv_url: str

    # these are defined in the inventory file itself
    name: str
    domain: str
    role: str
    priority: str
    uri: str
    dispname: str

    @property
    def full_uri(self):
        # TODO: this should only apply to a uri ending with "$"
        return self.inv_url + self.uri.replace("$", self.name)

    @classmethod
    def make_simple(cls, inv_name, full_url, name, role="function"):
        return cls(
            inv_name,
            "",
            name,
            domain="py",
            role=role,
            priority=1,
            uri=full_url,
            dispname="-",
        )


class Inventories:
    def __init__(self):
        self.registry: dict[str, list[EnhancedItem]] = {}

    def items(self):
        return itertools.chain(*self.registry.values())

    def load_inventory(self, inventory: dict, url: str, invname: str):
        all_items = []
        for item in inventory["items"]:
            # TODO: what are the rules for inventories with overlapping names?
            #       it seems like this is where priority and using source name as an
            #       optional prefix in references is useful (e.g. siuba:a.b.c).
            enh_item = EnhancedItem(inv_name=invname, inv_url=url, **item)
            all_items.append(enh_item)

        self.registry[invname] = all_items

    def lookup_reference(self, ref: Ref) -> EnhancedItem:
        """Return the item corresponding to a reference."""

        crnt_items = self.items()
        for field in ["name", "role", "domain", "invname"]:
            if field == "name":
                # target may have ~ option in front, so we strip it off
                field_value = ref.target.lstrip("~")
            else:
                field_value = getattr(ref, field)

            if field == "role":
                # for some reason, things like :func: are short for :function:.
                field_value = self.normalize_role(field_value)

            crnt_items = self._filter_by_field(crnt_items, field, field_value)

        results = list(crnt_items)
        if not results:
            raise InvLookupError(
                f"Cross reference not found in an inventory file: `{ref}`"
            )

        if len(results) > 1:
            raise InvLookupError(
                f"Cross reference matches multiple entries.\n"
                f"Matching entries: {len(results)}\n"
                f"Reference: {ref}\n"
                f"Top 2 matches: \n  * {results[0]}\n  * {results[1]}"
            )

        return results[0]

    def normalize_role(self, role_name):
        """Normalize the role portion of a reference."""

        if role_name == "func":
            return "function"

        return role_name

    def ref_to_anchor(self, ref: str | Ref, text: "str | None"):
        """Return a Link element based on a reference in interlink format

        Parameters
        ----------
        ref:
            The interlink reference (e.g. "my_module.my_function").
        text:
            The text to be displayed for the link.

        Examples
        --------

        >>> url = "https://example.org/functools.partial.html"
        >>> item = EnhancedItem.make_simple('someinv', url, name = 'functools.partial')
        >>> invs = Inventories.from_items([item])
        >>> invs.ref_to_anchor("functools.partial")
        Link(content='functools.partial', url='https://example.org/functools.partial.html')

        >>> invs.ref_to_anchor("~functools.partial")
        Link(content='partial', url='https://example.org/functools.partial.html')
        """

        if isinstance(ref, str):
            ref = Ref.from_string(ref)

        is_shortened = ref.target.startswith("~")

        entry = self.lookup_reference(ref)
        dst_url = entry.full_uri

        if not text:
            name = entry.name if entry.dispname == "-" else entry.dispname
            if is_shortened:
                # shorten names from module.sub_module.func_name -> func_name
                name = name.split(".")[-1]
            return Link(content=name, url=dst_url)

        return Link(content=text, url=dst_url)

    def pandoc_ref_to_anchor(self, ref: str, text: str) -> Link | Code | Unchanged:
        """Convert a ref to a Link, with special handling for pandoc filters.

        Note that this function is similar to ref_to_anchor, but handles pandoc's
        representation of ` as "%60", uses warnings instead of errors, and returns
        non-ref urls unchanged.
        """

        # detect what *might* be an interlink. note that we don't validate
        # that it has a closing `, to allow a RefSyntaxError to bubble up.
        if ref.startswith("%60") or ref.startswith(":"):
            # Get URL ----
            try:
                return self.ref_to_anchor(ref.replace("%60", "`"), text)
            except InvLookupError as e:
                warnings.warn(f"{e.__class__.__name__}: {e}")
                if text:
                    # Assuming content is a ListContainer(Str(...))
                    body = text
                else:
                    body = ref.replace("%60", "`")
                return Code(content=body)

        return Unchanged(content=ref)

    @staticmethod
    def _filter_by_field(items, field_name: str, value: "str | None" = None):
        if value is None:
            return items

        # TODO: Ref uses invname, while EnhancedItem uses inv_name
        if field_name == "invname":
            field_name = "inv_name"

        return (item for item in items if getattr(item, field_name) == value)

    @classmethod
    def from_items(cls, items: "list[EnhancedItem]"):
        invs = cls()
        for item in items:
            items = invs.registry.setdefault(item.inv_name, [])
            items.append(item)

        return invs

    @classmethod
    def from_quarto_config(cls, cfg: str | dict, root_dir: str | None = None):
        if isinstance(cfg, str):
            if root_dir is None:
                root_dir = Path(cfg).parent

            cfg = yaml.safe_load(open(cfg))

        invs = cls()
        p_root = get_path_to_root() if root_dir is None else Path(root_dir)

        interlinks = cfg["interlinks"]
        sources = interlinks["sources"]
        cache = interlinks.get("cache", "_inv")

        # load this sites inventory ----
        site_inv = interlinks.get("site_inv", "objects.json")

        json_data = json.load(open(p_root / site_inv))
        invs.load_inventory(json_data, url="/", invname="")

        # load other inventories ----
        for doc_name, cfg in sources.items():
            fname = doc_name + "_objects.json"
            inv_path = p_root / Path(cache) / fname

            json_data = json.load(open(inv_path))

            invs.load_inventory(json_data, url=cfg["url"], invname=doc_name)

        return invs
