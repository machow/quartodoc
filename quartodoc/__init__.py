"""quartodoc is a package for building delightful python API documentation.
"""

# flake8: noqa

from .autosummary import get_function, get_object, Builder
from .renderers import MdRenderer
from .inventory import convert_inventory, create_inventory
from .ast import preview
from .builder.blueprint import blueprint
from .builder.collect import collect
from .layout import Auto

__all__ = (
    "Auto",
    "blueprint",
    "collect",
    "convert_inventory",
    "create_inventory",
    "get_object",
    "preview",
    "Builder",
    "BuilderPkgdown",
    "BuilderSinglePage",
    "MdRenderer",
)
