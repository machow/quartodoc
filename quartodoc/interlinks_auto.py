from __future__ import annotations

import jedi

import ast
from fastapi import FastAPI
from pydantic import BaseModel
import uvicorn

from typing_extensions import TypeAlias, Union

ASTItems: TypeAlias = Union[ast.Call, ast.Name, ast.Attribute]


class CallVisitor(ast.NodeVisitor):
    results: list[ASTItems]

    def __init__(self):
        self.results = []

    def visit_Call(self, node):
        if isinstance(node.func, ast.Attribute):
            self.results.append(node)
            return self.visit(node.func.value)
        elif isinstance(node.func, ast.Name):
            self.results.append(node)

    def visit_Name(self, node):
        self.results.append(node)

    def visit_Attribute(self, node):
        self.results.append(node)

    def reset(self):
        self.results = []

        return self

    @classmethod
    def analyze(cls, code) -> list[Interlink]:
        visitor = cls()
        visitor.visit(ast.parse(code))
        script = jedi.Script(code)
        all_links = [node_to_interlink_name(script, call) for call in visitor.results]
        return [link for link in all_links if link is not None]


def narrow_node_start(node: ast.AST) -> tuple[int, int]:
    if isinstance(node, ast.Attribute):
        return (node.value.end_lineno, node.value.end_col_offset)

    return node.lineno, node.col_offset


def node_to_interlink_name(script: jedi.Script, node: ast.AST) -> Interlink | None:
    print(node.func.lineno, node.func.end_col_offset)
    try:
        func = node.func
        name = script.goto(func.lineno, func.end_col_offset)[0]
        full_name = name.full_name
        lineno, col_offset = narrow_node_start(func)
        return Interlink(
            name=full_name,
            lineno=lineno,
            end_lineno=func.end_lineno,
            col_offset=col_offset,
            end_col_offset=func.end_col_offset,
        )
    except IndexError:
        return None


class Code(BaseModel):
    content: str


class Interlink(BaseModel):
    name: str
    lineno: int
    end_lineno: int
    col_offset: int
    end_col_offset: int


app = FastAPI()


@app.post("/analyze")
async def analyze(code: Code):
    res = CallVisitor.analyze(code.content)
    return {"interlinks": res}


if __name__ == "__main__":
    import sys

    port = int(sys.argv[1])
    uvicorn.run(app, port=port, host="0.0.0.0")
