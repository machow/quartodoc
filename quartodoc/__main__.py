import click
import contextlib
import os
import time
import sphobjinv as soi
import yaml
import importlib
from pathlib import Path
from watchdog.observers import Observer
from functools import partial
from watchdog.events import PatternMatchingEventHandler
from quartodoc import Builder, convert_inventory
from pydantic import BaseModel


def get_package_path(package_name):
    """
    Get the path to a package installed in the current environment.
    """
    try:
        lib = importlib.import_module(package_name)
        return lib.__path__[0]
    except ModuleNotFoundError:
        raise ModuleNotFoundError(
            f"Package {package_name} not found.  Please install it in your environment."
        )


class FileInfo(BaseModel):
    size: int
    mtime: float
    name: str = ""


class QuartoDocFileChangeHandler(PatternMatchingEventHandler):
    """
    A handler for file changes.
    """

    # Ignore patterns for the file watcher that are not relevant to the docs
    py_ignore_patterns = [
        "*/__pycache__/*",  # These are the compiled python code files which are automatically generated by Python
        "*/.ipynb_checkpoints/*",  # This directory is created by Jupyter Notebook for auto-saving notebooks
        "*/.vscode/*",  # If you're using Visual Studio Code, it creates this directory to store settings specific to that project.
        "*/.idea/*",  # Similar to .vscode/, but for JetBrains IDEs like PyCharm.
        "*/.git/*",  # i This directory is created by Git. It's not relevant to the docs.
        "*/venv/*",
        "*/env/*",
        "*/.env/*",  # Common names for directories containing a Python virtual environment.
        "*/.pytest_cache/*",  # This directory is created when you run Pytest.
        "*/.eggs/*",
        "*/dist/*",
        "*/build/*",
        "*/*.egg-info/*",  # These are typically created when building Python packages with setuptools.
        "*.pyo",  # These are optimized .pyc files, created when Python is run with the -O flag.
        "*.pyd",  # This is the equivalent of a .pyc file, but for C extensions on Windows.
        "*/.mypy_cache/*",  # This directory is created when you run mypy.
    ]

    def __init__(self, callback):
        super().__init__(
            ignore_patterns=self.py_ignore_patterns, ignore_directories=True
        )
        self.callback = callback
        self.old_file_info = FileInfo(size=-1, mtime=-1, name="")

    def get_file_info(self, path: str) -> FileInfo:
        """
        Get the file size and modification time.
        """
        return FileInfo(
            size=os.stat(path).st_size, mtime=os.stat(path).st_mtime, name=path
        )

    def is_diff(self, old: FileInfo, new: FileInfo) -> bool:
        """
        Check if a file has changed. Prevents duplicate events from being triggered.
        """
        same_nm = old.name == new.name
        diff_sz = old.size != new.size
        diff_tm = new.mtime - old.mtime  # wait 1/4 second before triggering

        if (
            diff_tm < 0.25
        ):  # if consequetive events are less than 1/4th of a second apart, ignore
            return False
        elif same_nm:
            if diff_sz or diff_tm >= 0.25:
                return True
            else:
                return False
        else:
            return True

    def callback_if_diff(self, event):
        """
        Call the callback if the file has changed.
        """
        new_file_info = self.get_file_info(event.src_path)
        if self.is_diff(self.old_file_info, new_file_info):
            self.print_event(event)
            self.callback()
        self.old_file_info = new_file_info

    @classmethod
    def print_event(cls, event):
        print(f"Rebuilding docs.  Detected: {event.event_type} path : {event.src_path}")

    def on_modified(self, event):
        self.callback_if_diff(event)

    def on_created(self, event):
        self.callback_if_diff(event)


def _enable_logs():
    import logging
    import sys

    root = logging.getLogger("quartodoc")
    root.setLevel(logging.INFO)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )
    handler.setFormatter(formatter)
    root.addHandler(handler)


@contextlib.contextmanager
def chdir(new_dir):
    prev = os.getcwd()
    os.chdir(new_dir)
    try:
        yield new_dir
    finally:
        os.chdir(prev)


@click.group()
def cli():
    pass


@click.command()
@click.option(
    "--config",
    default="_quarto.yml",
    help="Change the path to the configuration file.  The default is `./_quarto.yml`",
)
@click.option(
    "--filter",
    nargs=1,
    default="*",
    help="Specify the filter to select specific files. The default is '*' which selects all files.",
)
@click.option(
    "--dry-run",
    is_flag=True,
    default=False,
    help="If set, prevents new documents from being generated.",
)
@click.option(
    "--watch",
    is_flag=True,
    default=False,
    help="If set, the command will keep running and watch for changes in the package directory.",
)
@click.option("--verbose", is_flag=True, default=False, help="Enable verbose logging.")
def build(config, filter, dry_run, watch, verbose):
    """
    Generate API docs based on the given configuration file  (`./_quarto.yml` by default).
    """
    cfg_path = f"{os.getcwd()}/{config}"
    if not Path(cfg_path).exists():
        raise FileNotFoundError(
            f"Configuration file {cfg_path} not found.  Please create one."
        )
    if verbose:
        _enable_logs()

    builder = Builder.from_quarto_config(config)
    doc_build = partial(builder.build, filter=filter)

    if dry_run:
        pass
    else:
        with chdir(Path(config).parent):
            if watch:
                pkg_path = get_package_path(builder.package)
                print(f"Watching {pkg_path} for changes...")
                observer = Observer()
                observer._event_queue.maxsize = 1  # the default is 0 which is infinite, and there isn't a way to set this in the constructor
                event_handler = QuartoDocFileChangeHandler(callback=doc_build)
                observer.schedule(event_handler, pkg_path, recursive=True)
                observer.schedule(event_handler, cfg_path, recursive=True)
                observer.start()
                try:
                    while True:
                        time.sleep(1)
                except KeyboardInterrupt:
                    pass
                finally:
                    observer.stop()
                    observer.join()
            else:
                doc_build()


@click.command()
@click.argument("config", default="_quarto.yml")
@click.option("--dry-run", is_flag=True, default=False)
def interlinks(config, dry_run):
    cfg = yaml.safe_load(open(config))
    interlinks = cfg.get("interlinks", None)

    cache = cfg.get("cache", "_inv")

    p_root = Path(config).parent

    if interlinks is None:
        print("No interlinks field found in your quarto config. Quitting.")
        return

    for k, v in interlinks["sources"].items():
        # TODO: user shouldn't need to include their own docs in interlinks
        if v["url"] == "/":
            continue

        url = v["url"] + v.get("inv", "objects.inv")
        inv = soi.Inventory(url=url)

        p_dst = p_root / cache / f"{k}_objects.json"
        p_dst.parent.mkdir(exist_ok=True, parents=True)

        convert_inventory(inv, p_dst)


cli.add_command(build)
cli.add_command(interlinks)


if __name__ == "__main__":
    cli()
