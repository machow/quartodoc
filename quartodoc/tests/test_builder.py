import pytest
import yaml

from pathlib import Path
from quartodoc import layout as lo
from quartodoc import Builder, blueprint


@pytest.fixture
def builder(tmp_path):
    section = lo.Section(
        title="abc",
        desc="xyz",
        contents=[
            lo.Auto(name="get_object"),
            lo.Auto(
                name="MdRenderer", members=["render", "summarize"], children="separate"
            ),
        ],
    )

    builder = Builder(package="quartodoc", sections=[section], dir=str(tmp_path))

    yield builder


def test_builder_build_filter_simple(builder):
    builder.build(filter="get_object")

    assert (Path(builder.dir) / "get_object.qmd").exists()
    assert not (Path(builder.dir) / "MdRenderer.qmd").exists()


def test_builder_build_filter_wildcard_class(builder):
    builder.build(filter="MdRenderer*")

    len(list(Path(builder.dir).glob("Mdrenderer*"))) == 3


def test_builder_build_filter_wildcard_methods(builder):
    builder.build(filter="MdRenderer.*")

    len(list(Path(builder.dir).glob("Mdrenderer.*"))) == 2


def test_builder_auto_options():
    cfg = yaml.safe_load(
        """
    quartodoc:
      package: quartodoc
      options:
        members: [a_func, a_attr]
      sections:
        - contents: [quartodoc.tests.example]
    """
    )

    builder = Builder.from_quarto_config(cfg)
    assert builder.layout.options.members == ["a_func", "a_attr"]


def test_builder_generate_sidebar(tmp_path, snapshot):
    cfg = yaml.safe_load(
        """
    quartodoc:
      package: quartodoc.tests.example
      sections:
        - title: first section
          desc: some description
          contents: [a_func]
        - title: second section
          desc: title description
        - subtitle: a subsection
          desc: subtitle description
          contents:
            - a_attr
    """
    )

    builder = Builder.from_quarto_config(cfg)
    bp = blueprint(builder.layout)
    d_sidebar = builder._generate_sidebar(bp)

    assert yaml.dump(d_sidebar) == snapshot
