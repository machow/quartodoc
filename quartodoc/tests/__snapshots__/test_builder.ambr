# serializer version: 1
# name: test_builder_generate_sidebar
  '''
  website:
    sidebar:
    - contents:
      - reference/index.qmd
      - contents:
        - reference/a_func.qmd
        section: first section
      - contents:
        - contents:
          - reference/a_attr.qmd
          section: a subsection
        section: second section
      id: reference
    - id: dummy-sidebar
  
  '''
# ---
